packages/discovery-provider/plugins/pedalboard/apps/solana-relay/src/routes/listen/trackListenInstructions.ts

```diff
@@ -0,0 +1,213 @@
+        this.trackData = trackData
+        this.signature = signature
+        this.recoveryId = recoveryId
+    }
+}
+
+class InstructionEnum {
+    instruction: InstructionArgs;
+    choose: string;
+    constructor({ instruction, choose }: { instruction: InstructionArgs, choose: string }) {
+        this.instruction = instruction
+        this.choose = choose
+    }
+}
+
+// @ts-expect-error instructionSchema doesn't have correct type when converted from js
+const instructionSchema = new Map([
+    [
+        InstructionEnum,
+        {
+            kind: 'enum',
+            field: 'choose',
+            values: [['instruction', InstructionArgs]]
+        }
+    ],
+    [
+        InstructionArgs,
+        {
+            kind: 'struct',
+            fields: [
+                ['trackData', TrackData],
+                ['signature', [64]],
+                ['recoveryId', 'u8']
+            ]
+        }
+    ],
+    [
+        TrackData,
+        {
+            kind: 'struct',
+            fields: [
+                ['userId', 'string'],
+                ['trackId', 'string'],
+                ['source', 'string'],
+                ['timestamp', 'u64']
+            ]
+        }
+    ]
+])
+
+let cachedListenBlocktime: number | null = null
+let lastRetrievedListenBlocktime: number | null = null
+
+async function getListenTimestamp(connection: Connection) {
+    const currentEpochInSec = Math.round(Date.now() / 1000)
+    if (
+        cachedListenBlocktime && lastRetrievedListenBlocktime &&
+        Math.abs(lastRetrievedListenBlocktime - currentEpochInSec) < 30
+    ) {
+        return cachedListenBlocktime
+    }
+
+    const slot = await connection.getSlot('finalized')
+    const blockTime = await connection.getBlockTime(slot)
+
+    // update cached values
+    cachedListenBlocktime = blockTime
+    lastRetrievedListenBlocktime = currentEpochInSec
+
+    return blockTime
+}
+
+export function getFeePayerKeyPair(): Keypair {
+    const feePayers = config.solanaFeePayerWallets
+    const randomFeePayerIndex = Math.floor(
+        Math.random() * feePayers.length
+    )
+    return feePayers[randomFeePayerIndex]
+}
+
+export async function createTrackListenInstructions({
+    logger,
+    userId,
+    trackId,
+    location
+}: { logger: Logger,userId: string, trackId: string, location: LocationData }): Promise<TransactionInstruction[]> {
+    const trackListenProgram = new PublicKey(config.trackListenCountProgramId)
+    const ethRegistryProgram = new PublicKey(config.ethRegistryProgramId)
+    const validSignerPubK = new PublicKey(config.listensValidSigner)
+    const privKey = Buffer.from(config.solanaSignerPrivateKey, 'hex')
+    const pubKey = secp256k1.publicKeyCreate(privKey, false).slice(1)
+
+    logger.info({ trackListenProgram, ethRegistryProgram, validSignerPubK }, "ids and keys")
+
+    const connection = connections[0]
+    const source = 'relay'
```

Comment:
do we want more info in this? like the node addr / endpoint or something?

Reply:
What was it before?

Reply:
it was just the string 'relay' before but i assume that's because we only had one relay entrypoint? we'd have a few more now, i'm not actually the full use of this value
src: https://github.com/AudiusProject/audius-protocol/blob/main/packages/identity-service/src/routes/trackListens.js#L449